src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+                   action='append')
+
+    p.add_argument('--output', '-o',
+                   nargs='?')
+
+    p.add_argument('--key',
+                   type=pathlib.Path)
+    p.add_argument('--cert',
+                   type=pathlib.Path)
+
+    p.add_argument('--measure',
+                   action='store_true')
+
+    args = p.parse_args()
+
+    if bool(args.key) ^ bool(args.cert):
+        raise ValueError('--key and --cert must be specified together')
+
+    if args.os_release is None:
+        p = pathlib.Path('/etc/os-release')
+        if not p.exists():
+            p = pathlib.Path('/usr/lib/os-release')
+        args.os_release = p
+
+    if args.efi_arch is None:
+        args.efi_arch = guess_efi_arch()
+
+    if args.stub is None:
+        args.stub = f'/usr/lib/systemd/boot/efi/linux{args.efi_arch}.efi.stub'
+
+    if args.output is None:
+        suffix = '.efi' if args.key else '.unsigned.efi'
+        args.output = args.linux.name + suffix
+
+    return args
+
+
+def check_inputs(args):
+    for name, value in vars(args).items():
+        if name == 'output':
+            continue
+
+        if not isinstance(value, pathlib.Path):
+            continue
+
+        # Open file to check that we can read it, or generate an exception
+        open(value).close()
+
+
+def make_uki(args):
+    sections = [
+        # FIXME: calculate offsets based on content size
+        ('osrel',   args.os_release,    0x20_000),
+        ('cmdline', args.cmdline,       0x30_000),
+        ('dtb',     args.devicetree,    0x40_000),
+        ('splash',  args.splash,       0x100_000),
+        ('initrd',  args.initrd,     0x3_000_000),
+        # linux shall be last to leave breathing room for decompression
+        ('linux',   args.linux,      0x2_000_000),
+    ]
+
+    sections = [Section.create(name, content, offset)
+                for name, content, offset in sections
+                if content]
+
+    # Insert extra sections just before .linux
+    objcopy_sections = sections[:-1] + args.sections + sections[-1:]
+
+    if args.key:
+        unsigned = tempfile.NamedTemporaryFile(prefix='linux')
+        output = unsigned.name
+    else:
+        output = args.output
+
+    objcopy = [
+        'objcopy',
+        args.stub,
+        *itertools.chain.from_iterable(
+            ('--add-section',        f'.{s.name}={s.content}',
+             '--change-section-vma', f'.{s.name}=0x{s.offset:x}')
+            for s in objcopy_sections),
+        output,
+    ]
+    print('+', shell_join(objcopy))
+    subprocess.check_call(objcopy)
+
+    # signing
+
+    if args.key:
+        sbsign = [
+            'sbsign',
+            '--key', args.key,
+            '--cert', args.cert,
+            unsigned.name,
+            '--output', args.output,
+        ]
+        print('+', shell_join(sbsign))
+        subprocess.check_call(sbsign)
+
+        # We end up with no executable bits, let's reapply them
+        os.umask(umask := os.umask(0))
```

Comment:
The walrus operator was introduced with Python 3.8. I don't know what the lower Python version requirement is, but if it's 3.8 you can use `shlex.join` as mentioned above.